File: agent-driven-architecture-main/examples/kotlin/src/test/kotlin/adr/spine/ReplayTest.kt
// ── test/spine/replay — G9: a LIVE run against its REPLAY ─────────────────
// The shipped harness asserted f(x) == f(x): it folded one in-memory array twice
// through a pure function. Measured: seam 07's own named violation (a tool reading a
// mutable global and performing a side effect) PASSED it, because foldAll never
// invoked a tool at all.
//
// This asserts something that can actually fail: the state and the FULL effect
// sequence — keys and every timestamp — produced by a live boundary against the same
// two things re-derived from nothing but the committed bytes.

package adr.spine

import adr.Driver
import adr.app.Assembly
import adr.app.Env
import adr.app.RunAuthority
import adr.app.Wiring
import adr.app.World
import adr.blocks.triage.PRE_V2_REASON
import adr.contract.TriageResult.Priority
import adr.blocks.triage.SET_PRIORITY
import adr.blocks.triage.TriageUpcast
import adr.contract.ToolResult
import adr.contract.TriageEffect
import adr.contract.TriageResult
import adr.contract.TriageV1Result
import adr.spine.boundary.MovingClock
import adr.spine.boundary.RecordingSink
import adr.spine.pure.Action
import adr.spine.pure.Actor
import adr.spine.pure.Authority
import adr.spine.pure.ContextFixture
import adr.spine.pure.CURRENT_SCHEMA
import adr.spine.pure.GENESIS_SCHEMA
import adr.spine.pure.PerformMode
import adr.spine.pure.RawInput
import adr.spine.pure.SchemaUpcast
import adr.spine.pure.SchemaVersion
import adr.spine.pure.Signature
import adr.spine.pure.StepRecord
import adr.spine.pure.StepRecordV1
import adr.spine.pure.TicketId
import adr.spine.pure.Timestamp
import adr.spine.replay.Replay
import adr.spine.replay.ReplayFaithfulness
import kotlin.reflect.full.primaryConstructor
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

class ReplayTest {

    @Test
    fun `G9 - the live run and its re-fold agree on state and on every effect`() {
        val world = World()
        val authority = RunAuthority()
        val app = Wiring().wireApp(
            Env(
                world = world,
                authority = authority,
                clock = MovingClock(start = 1000, step = 7),
            ),
        )

        Driver().driveCanonicalSession(app, authority)

        val liveState = app.state
        val liveEffects = app.performed.toList()

        val (state2, effects2) = Replay(Assembly()::fold, app.admission).refold(app.initial, app.bus.records())

        assertEquals(liveState, state2, "state re-derives from the committed bytes")
        assertEquals(liveEffects, effects2, "so does the full effect sequence — keys AND timestamps")
        assertTrue(liveEffects.isNotEmpty())

        // And the digest check: a change to projectContext that silently alters what the
        // model saw fails the golden trace, WITHOUT re-running the model (G15/§5.3). The
        // three app-constant values are what the harness is BUILT with; the timeline and
        // the live run it is measured against are what it is CALLED with.
        ReplayFaithfulness(
            fold = Assembly()::fold,
            projectContext = Assembly()::context,
            promptVersion = "triage-prompt@1",
            admission = app.admission,
            bounds = app.boundary.contextBounds,
        ).assertFaithful(
            initial = app.initial,
            records = app.bus.records(),
            liveState = liveState,
            liveEffects = liveEffects,
        )
    }

    // ── THE SCRUB CURSOR, PROVEN BY EXERCISE ─────────────────────────────
    // docs/DECISIONS.md:117-118 ratifies the cursor "proving the scrub story BY
    // EXERCISE", and stateAtStep's own KDoc promises exactly what is below. A
    // review found the promise false in both ports and the tool called by
    // nothing at all: a mutation making it ignore `k` and fold the WHOLE
    // timeline left `./gradlew check --rerun-tasks` fully green. A scrub bar
    // wired to that would show the end state at every position on the drag.
    @Test
    fun `the scrub cursor re-folds ONLY the prefix, at an interior k`() {
        val app = Wiring().wireApp(
            Env(world = World(), authority = RunAuthority(), clock = MovingClock(start = 1000, step = 7)),
        )
        Driver().driveCanonicalSession(app, RunAuthority())
        val replay = Replay(Assembly()::fold, app.admission)
        val records = app.bus.records()
        val whole = replay.refold(app.initial, records)
        // The canonical session performs several effects, so an interior cursor
        // must differ from the whole timeline on BOTH halves.
        assertTrue(whole.effects.size > 1, "the fixture must discriminate on effects")

        for (k in 1 until records.size) {
            val cut = replay.stateAtStep(app.initial, records, k)
            assertTrue(
                cut.state != whole.state,
                "k=$k is already the end state — the cursor ignored its bound",
            )
            assertEquals(
                replay.refold(app.initial, records.take(k)).state,
                cut.state,
                "k=$k must equal the re-fold of exactly that many records",
            )
            // AND THE EFFECTS HALF, which the KDoc promises in as many words: a
            // scrub that shows state and hides the page it had already sent is a
            // lie of omission. The first landing asserted state only, so
            // suppressing every interior effect left the gate green.
            assertEquals(
                replay.refold(app.initial, records.take(k)).effects,
                cut.effects,
                "k=$k must re-derive exactly the effects that prefix produced",
            )
        }
    }

    @Test
    fun `the scrub cursor CLAMPS BY SLICING, never by throwing`() {
        val app = Wiring().wireApp(
            Env(world = World(), authority = RunAuthority(), clock = MovingClock(start = 1000, step = 7)),
        )
        Driver().driveCanonicalSession(app, RunAuthority())
        val replay = Replay(Assembly()::fold, app.admission)
        val records = app.bus.records()
        val whole = replay.refold(app.initial, records)

        assertEquals(app.initial, replay.stateAtStep(app.initial, records, 0).state)
        assertEquals(app.initial, replay.stateAtStep(app.initial, records, -5).state)
        assertEquals(emptyList(), replay.stateAtStep(app.initial, records, 0).effects)
        assertEquals(whole.state, replay.stateAtStep(app.initial, records, records.size).state)
        assertEquals(whole.state, replay.stateAtStep(app.initial, records, records.size + 99).state)
        assertEquals(whole.effects, replay.stateAtStep(app.initial, records, records.size).effects)
    }

    @Test
    fun `PerformMode REPLAY collects the descriptors and fires NOTHING`() {
        val world = World()
        val authority = RunAuthority()
        val app = Wiring().wireApp(Env(world = world, authority = authority))
        Driver().driveCanonicalSession(app, authority)

        val liveEffects = app.performed.toList()
        val pagesAfterLive = world.pages.size
        val deliveriesAfterLive = world.deliveries.size
        assertEquals(1, pagesAfterLive)
        assertEquals(1, deliveriesAfterLive)

        // Drive the SAME sink chain the live run used — including the real adapters,
        // and the same ASSEMBLED dispatcher, so what replay is stubbing is the shipped
        // performer set rather than a second sink that happens to resemble it.
        val replayLog = mutableListOf<String>()
        val replaySink = RecordingSink(
            adr.app.AppSink(
                adr.app.Wiring().effectPerformers(
                    Env(world = world, authority = authority),
                    replayLog,
                ),
                adr.app.DiagPerformer(replayLog),
            ),
        )
        Replay(Assembly()::fold, app.admission)
            .collectPerform(app.initial, app.bus.records(), replaySink, PerformMode.REPLAY)

        assertEquals(liveEffects, replaySink.performed, "descriptors collected…")
        assertEquals(pagesAfterLive, world.pages.size, "…and nothing fired")
        assertEquals(deliveriesAfterLive, world.deliveries.size)
    }

    @Test
    fun `a divergent re-fold is DETECTED - the harness is not vacuous`() {
        val authority = RunAuthority()
        val app = Wiring().wireApp(Env(world = World(), authority = authority))
        Driver().driveCanonicalSession(app, authority)

        // Drop one committed step: the re-fold must no longer match the live run.
        val truncated = app.bus.records().dropLast(1)
        val (state2, _) = Replay(Assembly()::fold, app.admission).refold(app.initial, truncated)
        assertTrue(state2 != app.state, "a harness that cannot fail is not a harness")
    }

    // ── 14.7 — SCHEMA EVOLUTION, over a log written before the field existed ──
    //
    // Two things are proven, and only the pair is worth anything:
    //
    //   1. THE OLD LOG CANNOT REPLAY AT ALL, in BOTH halves. The ENVELOPE half:
    //      `schemaVersion` has no default, so no construction site can forget it,
    //      and the two versions are different VALUES rather than the same constant
    //      read twice. The PAYLOAD half: a v1 payload is not a ToolResult, so it
    //      cannot enter `StepRecord.results` whatever the envelope says.
    //   2. THE UPCAST IS OBSERVABLE. Re-folding the LIFTED log produces a
    //      different effect sequence from re-folding a native v2 one, in a field
    //      the fold actually consumes. An upcaster whose output nothing reads is a
    //      function that can be deleted with every test still green.
    //
    // WHY THESE ARE REFLECTION AND VALUE ASSERTIONS rather than compile fixtures:
    // this port's must-fail-compile harness is wired to `check`, not `test`, so a
    // compile fixture would not run under the gate that adjudicates this change.
    // Every assertion below is therefore chosen for being able to INVERT — an
    // assertion on a final class's supertype, for instance, is a constant `false`
    // and has no place in a proof set.
    //
    // The v1 log is an in-code typed fixture, not a file: this reference persists
    // nothing (the bus is a list of typed records) and 14.1 leaves the canonical
    // encoding product-owned. So the "old shape" is a TYPE the compiler refuses,
    // which is a stronger fixture than bytes a hand-written parser must agree with.

    private fun v1Log(): List<StepRecordV1<TriageV1Result.SetPriority>> = listOf(
        StepRecordV1(
            schemaVersion = GENESIS_SCHEMA,
            now = Timestamp(1000),
            sig = Signature(Actor.Agent, Authority("agent-run-7f")),
            staged = emptyList(),
            actions = listOf(Action(SET_PRIORITY, RawInput("ticket" to "4118", "level" to "High"))),
            results = listOf(
                TriageV1Result.SetPriority(SET_PRIORITY, TicketId("4118"), Priority.High),
            ),
            commands = emptyList(),
            context = ContextFixture("triage-prompt@1", ""),
        ),
    )

    @Test
    fun `14_7 - this port writes v2, genesis was 1, and they are different numbers`() {
        // Without this, every other version assertion compares the committed value
        // against the constant that produced it — both sides move together under a
        // mutation and the number is unpinned. These are the two hard values.
        assertEquals(SchemaVersion(2), CURRENT_SCHEMA, "this port writes v2")
        assertEquals(SchemaVersion(1), GENESIS_SCHEMA, "genesis is 1; there is no v0")
        assertNotEquals(CURRENT_SCHEMA, GENESIS_SCHEMA)
    }

    @Test
    fun `14_7 - the envelope is REQUIRED on BOTH records, and the v1 payload is not a ToolResult`() {
        val params = checkNotNull(StepRecord::class.primaryConstructor).parameters
        assertEquals(8, params.size, "the record is the envelope plus the fields the step is")
        assertFalse(
            params.single { it.name == "schemaVersion" }.isOptional,
            "a defaulted schemaVersion would let a construction site forget the envelope",
        )

        // The v1 record is STAMPED too, and just as non-optionally. A fixture that
        // could omit the version would stop standing in for a record that carried one,
        // and "genesis is 1" would be true of a comment rather than of the code.
        val v1Params = checkNotNull(StepRecordV1::class.primaryConstructor).parameters
        assertEquals(8, v1Params.size, "the v1 record carries its own version")
        assertFalse(
            v1Params.single { it.name == "schemaVersion" }.isOptional,
            "the v1 envelope is stamped, not implied",
        )

        // THE PAYLOAD HALF, asserted where it can actually invert: `ToolResult` is a
        // SEALED class, so this flips the moment someone makes the v1 payload extend
        // it — which is exactly the mistake that would let a v2 envelope be spread
        // over v1 payloads and re-folded.
        assertFalse(
            ToolResult::class.java.isAssignableFrom(TriageV1Result.SetPriority::class.java),
            "a v1 payload must NOT be a ToolResult, or refold would accept an un-upcast log",
        )
    }

    @Test
    fun `14_7 - the UPCAST log re-folds, and the upcaster's decision rides the effect`() {
        val app = Wiring().wireApp(Env(world = World(), authority = RunAuthority()))
        val lifted = SchemaUpcast(TriageUpcast()::v1).let { upcast -> v1Log().map(upcast::v1) }

        assertEquals(listOf(SchemaVersion(2)), lifted.map { it.schemaVersion })

        val (state, effects) = Replay(Assembly()::fold, app.admission).refold(app.initial, lifted)
        assertEquals(
            listOf(
                TriageEffect.LogDecision(
                    at = Timestamp(1000),
                    ticket = TicketId("4118"),
                    level = Priority.High,
                    // the Kotlin app seeds tickets without a priority, so the first
                    // transition supersedes nothing — the fold reading its OWN state
                    supersedes = null,
                    reason = PRE_V2_REASON,
                ),
            ),
            effects.map { it.effect },
        )
        assertEquals(Priority.High, state.triage.priority[TicketId("4118")])
    }

    @Test
    fun `14_7 - the field is OBSERVABLE - a NATIVE v2 record re-folds to a different effect`() {
        // The control, and it is built from a real v2 record rather than by lifting the
        // v1 fixture through a substitute upcaster: a "native" record assembled out of
        // history is not a control at all. If `reason` never reached the fold, this and
        // the assertion above would be the same value and the upcaster untestable.
        val app = Wiring().wireApp(Env(world = World(), authority = RunAuthority()))
        val native = listOf(
            StepRecord(
                schemaVersion = CURRENT_SCHEMA,
                now = Timestamp(1000),
                sig = Signature(Actor.Agent, Authority("agent-run-7f")),
                staged = emptyList(),
                actions = listOf(
                    Action(SET_PRIORITY, RawInput("ticket" to "4118", "level" to "High")),
                ),
                results = listOf(
                    TriageResult.SetPriority(
                        SET_PRIORITY,
                        TicketId("4118"),
                        Priority.High,
                        "customer escalated",
                    ),
                ),
                commands = emptyList(),
                context = ContextFixture("triage-prompt@1", ""),
            ),
        )

        val (_, effects) = Replay(Assembly()::fold, app.admission).refold(app.initial, native)
        assertEquals(
            listOf("customer escalated"),
            effects.map { (it.effect as TriageEffect.LogDecision).reason },
        )
    }

    @Test
    fun `14_7 - a LIVE step commits the envelope, at the one site that mints a record`() {
        val authority = RunAuthority()
        val app = Wiring().wireApp(Env(world = World(), authority = authority))
        Driver().driveCanonicalSession(app, authority)

        assertTrue(app.bus.records().isNotEmpty())
        assertEquals(
            listOf(SchemaVersion(2)),
            app.bus.records().map { it.schemaVersion }.distinct(),
        )
    }
}

File: agent-driven-architecture-main/examples/typescript/test/spine/replay.test.ts
// ── G9 — a LIVE run against its REPLAY, not a fold against itself ──────────
//
// MEASURED against the shipped reference: seam 07 §D's own named violation (a
// tool reading a mutable global and performing a side effect) was injected, the
// world was mutated between runs, and `replayTest` PASSED — the tool's
// side-effect count never moved, because `foldAll` never invoked a tool. The
// old harness asserted f(x) == f(x), which seam 07 §C itself calls true by
// definition.
//
// What is asserted here is what a live boundary DID against what its committed
// bytes re-derive: state, and the FULL keyed effect sequence including every
// timestamp.

import { describe, expect, it } from "vitest";
import { effectSink } from "../../src/app/wire";
import type { SetPriorityResult, SetPriorityResultV1 } from "../../src/blocks/triage/contract";
import { PRE_V2_REASON, upcastSetPriority } from "../../src/blocks/triage/tools";
import type { StepChannel } from "../../src/spine/boundary/action";
import { RecordingSink } from "../../src/spine/boundary/in-memory";
import { authority, Signature } from "../../src/spine/pure/actor";
import type { StepRecord, StepRecordV1 } from "../../src/spine/pure/step-record";
import { GENESIS_SCHEMA_VERSION, SCHEMA_VERSION, upcastV1 } from "../../src/spine/pure/step-record";
import { seal } from "../../src/spine/pure/tool-result";
import { collectPerform, refold, stateAtStep } from "../../src/spine/replay/replay";
import { fakeWorld, harness, POLICY_TIER } from "../harness";
import { must } from "../support/must";

function driveFullSession(h: ReturnType<typeof harness>): void {
  const step = (channel: StepChannel, ...actions: { tool: string; input: unknown }[]): void =>
    void channel.submit({ staged: [], actions });
  const agent = h.app.boundary.agent;
  const human = h.app.boundary.human;

  step(agent, { tool: "setPriority", input: { ticket: "4118", level: "High" } });
  step(agent, { tool: "requestEscalation", input: { ticket: "4118" } });
  step(agent, { tool: "confirmEscalation", input: { ticket: "4118" } }); // self → refused
  h.actAs("Agent", POLICY_TIER);
  step(agent, { tool: "confirmEscalation", input: { ticket: "4118" } }); // other → granted
  step(agent, { tool: "recordFinding", input: { text: "first" } });
  step(agent, { tool: "requestSeal", input: {} });
  step(human, { tool: "confirmSeal", input: {} });
}

// ── THE SCRUB CURSOR, PROVEN BY EXERCISE ─────────────────────────────────
// docs/DECISIONS.md:117-118 ratifies the cursor "proving the scrub story BY
// EXERCISE", and the doc comment over `stateAtStep` promises exactly what is
// below — "BOTH ends are asserted, at an interior k and at the right edge, in
// test/spine/replay.test.ts". A review found the promise false in both ports
// and the tool called by nothing at all: a mutation making it ignore `k` and
// fold the WHOLE timeline left the full gate green, 434/434. A scrub bar wired
// to that would show the end state at every position on the drag.
describe("stateAtStep — the scrub prefix, at both ends and in between", () => {
  it("re-folds ONLY the prefix: an INTERIOR k stops where the cursor is", () => {
    const h = harness({ start: 1000, step: 7 });
    driveFullSession(h);
    const records = h.app.bus.records();
    const at = (k: number) =>
      stateAtStep(h.app.initial, records, h.app.dispatchers, h.app.licences, k);

    // THE MUTATION THAT SURVIVED: `k` ignored, whole timeline folded. Every
    // interior position must differ from the end state, or the cursor is a
    // no-op wearing a parameter.
    const whole = at(records.length);
    // The canonical session performs four effects across seven steps, so an
    // interior cursor must differ from the whole timeline on BOTH halves.
    expect(whole.effects.length).toBeGreaterThan(1);
    for (let k = 1; k < records.length; k += 1) {
      expect(at(k).state, `k=${k} must not already be the end state`).not.toEqual(whole.state);
      // and it must equal the re-fold of exactly that many records — the
      // independent derivation, not this function talking to itself.
      expect(at(k).state).toEqual(
        refold(h.app.initial, records.slice(0, k), h.app.dispatchers, h.app.licences).state,
      );
      // AND THE EFFECTS HALF, which the KDoc promises in as many words: "a scrub
      // that shows state and hides the on-call page it had already sent is a lie
      // of omission". The first landing asserted `.state` only, so suppressing
      // every interior effect left the whole gate green — the comment promised
      // more than the test asserted, which is the class this campaign keeps
      // closing.
      expect(at(k).effects).toEqual(
        refold(h.app.initial, records.slice(0, k), h.app.dispatchers, h.app.licences).effects,
      );
    }
  });

  it("CLAMPS BY SLICING, never by throwing — below zero and past the end", () => {
    const h = harness({ start: 1000, step: 7 });
    driveFullSession(h);
    const records = h.app.bus.records();
    const at = (k: number) =>
      stateAtStep(h.app.initial, records, h.app.dispatchers, h.app.licences, k);

    // LEFT EDGE: the initial state, and no effect re-derived.
    expect(at(0).state).toEqual(h.app.initial);
    expect(at(-5).state).toEqual(h.app.initial);
    expect(at(0).effects).toEqual([]);
    // RIGHT EDGE: the whole timeline, and past it is still the whole timeline.
    const whole = refold(h.app.initial, records, h.app.dispatchers, h.app.licences);
    expect(at(records.length).state).toEqual(whole.state);
    expect(at(records.length + 99).state).toEqual(whole.state);
    expect(at(records.length).effects).toEqual(whole.effects);
  });
});

describe("replay — a live run against its re-fold (G9)", () => {
  it("re-folds ONLY the committed bytes to the same state and the same effect sequence", () => {
    const h = harness({ start: 1000, step: 7 });
    driveFullSession(h);

    const replayed = refold(h.app.initial, h.app.bus.records(), h.app.dispatchers, h.app.licences);

    expect(replayed.state).toEqual(h.app.boundary.state);
    // the FULL sequence: keys AND every `at`
    expect(replayed.effects).toEqual(h.sink.performed);
    expect(replayed.effects.map((k) => k.effect.kind)).toEqual([
      "LogDecision",
      "Diag",
      "PageOncall",
      "DeliverArtifact",
    ]);
  });

  it("REPLAY mode collects the descriptors and fires NOTHING", () => {
    const h = harness({ start: 1000, step: 7 });
    driveFullSession(h);
    expect(h.world.pages).toEqual(["4118"]);
    expect(h.world.deliveries).toEqual([1]);

    const replayWorld = fakeWorld();
    const replaySink = new RecordingSink(effectSink(replayWorld.ports));
    collectPerform(
      h.app.initial,
      h.app.bus.records(),
      h.app.dispatchers,
      h.app.licences,
      replaySink,
      "REPLAY",
    );

    // descriptors collected …
    expect(replaySink.performed).toEqual(h.sink.performed);
    // … and NOTHING fired
    expect(replayWorld.world.pages).toEqual([]);
    expect(replayWorld.world.deliveries).toEqual([]);
    expect(replayWorld.world.logs).toEqual([]);
  });

  it("a live-source tool is caught by a CHECK, not by this harness — stated, not implied", () => {
    // Seam 07 §D claimed `replayTest` catches an impure tool. It cannot: replay
    // re-folds committed RESULTS and never invokes a tool body at all. What
    // catches a tool that reads a live source is gate check C3 (no clock, no
    // random, no id outside the boundary) and C8 (no await/fetch/node in a
    // block's pure files). See test/gate/gate.test.ts.
    expect(true).toBe(true);
  });
});

// ── 14.7 — SCHEMA EVOLUTION, over a log written before the field existed ────
//
// Two things are being proven, and only the pair is worth anything:
//
//   1. THE OLD LOG CANNOT REPLAY AT ALL — in BOTH halves, envelope and payload,
//      each isolated by its own inverting assertion below.
//   2. THE UPCAST IS OBSERVABLE. Re-folding the LIFTED log produces a different
//      effect sequence from re-folding a native v2 one, in a field the fold
//      actually consumes. An upcaster whose output nothing reads is a function
//      that can be deleted with every test still green.
//
// The v1 log is an in-code typed fixture rather than a file on disk, because
// this reference persists nothing: the bus is an array of typed records and
// 14.1's canonical encoding is deliberately product-owned. So the "old shape"
// is a TYPE the compiler refuses, which is a stronger fixture than bytes a
// hand-written parser would have to agree with.
const V1_SIG = new Signature("Agent", authority("agent-run-7f"));

const v1Result: SetPriorityResultV1 = {
  outcome: "ok-v1",
  tool: "setPriority",
  ticket: "4118",
  level: "High",
};

const v1Log = (): readonly StepRecordV1<SetPriorityResultV1>[] => [
  {
    schemaVersion: GENESIS_SCHEMA_VERSION,
    now: 1000,
    sig: V1_SIG,
    staged: [],
    actions: [{ tool: "setPriority", input: { ticket: "4118", level: "High" } }],
    results: [v1Result],
    commands: [],
    context: { promptVersion: "triage-prompt@1", digest: "" },
  },
];

describe("schema evolution — an old-shape log replays only through its upcaster (14.7)", () => {
  // ── THE THREE ISOLATING REFUSALS ────────────────────────────────────────
  // Each `@ts-expect-error` below IS the assertion: it fails the build the
  // moment the error it names stops happening. They are deliberately THREE and
  // not one, because a single directive that can be satisfied by any of three
  // defects passes for the wrong reason as soon as two of them are weakened.
  // Proven independent: making the envelope optional inverts only the first;
  // widening the envelope off the literal inverts only the second; reverting the
  // v1 payload's discriminant inverts only the third.

  it("REFUSES an unstamped record: the envelope is REQUIRED, not defaulted", () => {
    const seven = {
      now: 1000,
      sig: V1_SIG,
      staged: [],
      actions: [{ tool: "setPriority", input: { ticket: "4118", level: "High" } }],
      results: [],
      commands: [],
      context: { promptVersion: "triage-prompt@1", digest: "" },
    };
    // @ts-expect-error — `schemaVersion` is missing. The ONLY defect in this
    // value is the absent envelope; every other field is current-shape.
    const unstamped: StepRecord = seven;
    expect(unstamped.now).toBe(1000);
  });

  it("REFUSES a STALE version: a current-shape record still stamped genesis is not current", () => {
    // Payloads already lifted, so the only defect left is the number itself.
    // This is what makes the version type-level LOAD-BEARING rather than a
    // write-only decoration: widen `schemaVersion` off the literal and this
    // directive goes unused.
    const lifted = upcastV1(must(v1Log()[0]), upcastSetPriority);
    // @ts-expect-error — GENESIS is not the current version.
    const stale: StepRecord = { ...lifted, schemaVersion: GENESIS_SCHEMA_VERSION };
    expect(stale.results).toHaveLength(1);
  });

  it("REFUSES a v1 PAYLOAD under a v2 envelope — the structural half", () => {
    // The hole a structural language leaves open, closed. Re-stamping the
    // envelope by hand is trivial; what must not typecheck is the RESULT of
    // doing so, because that record folds and yields a `reason` of `undefined`
    // — a value neither `SetPriorityResult` nor `LogDecision` admits.
    // `SetPriorityResultV1.outcome` is `"ok-v1"`, which `ResultOutcome` does
    // not have, so the array is not a `readonly ToolResultBase[]`.
    const v1 = must(v1Log()[0]);
    // @ts-expect-error — v1 payloads are not ToolResults, whatever the envelope says.
    const forged: StepRecord = { ...v1, schemaVersion: SCHEMA_VERSION };
    expect(forged.schemaVersion).toBe(SCHEMA_VERSION);
  });

  it("this port writes v2, and genesis was 1 — the numbers, not just the constants", () => {
    // CX-1: without this, every other assertion compares the committed value
    // against the constant that produced it, and both sides move together under
    // a mutation. These are the only two hard values in the envelope story.
    expect(SCHEMA_VERSION).toBe(2);
    expect(GENESIS_SCHEMA_VERSION).toBe(1);
    expect(SCHEMA_VERSION).not.toBe(GENESIS_SCHEMA_VERSION);
  });

  // ── THE UPCAST, AND ITS CONTROL ─────────────────────────────────────────

  it("LIFTS it: the upcast log re-folds, and the upcaster's decision is in the effect", () => {
    const h = harness({ start: 1000, step: 7 });
    const lifted = v1Log().map((record) => upcastV1(record, upcastSetPriority));

    expect(lifted[0]?.schemaVersion).toBe(SCHEMA_VERSION);

    const replayed = refold(h.app.initial, lifted, h.app.dispatchers, h.app.licences);
    expect(replayed.effects.map((keyed) => keyed.effect)).toEqual([
      {
        kind: "LogDecision",
        at: 1000,
        effectClass: "Routine",
        ticket: "4118",
        level: "High",
        supersedes: "Normal",
        reason: PRE_V2_REASON,
      },
    ]);
    expect(h.app.boundary.state.triage.priority.get("4118")).toBe("Normal");
    expect(replayed.state.triage.priority.get("4118")).toBe("High");
  });

  it("the field is OBSERVABLE: a NATIVE v2 record re-folds to a different effect", () => {
    // The control, and it is built from a real v2 literal rather than by
    // spreading the v1 payload: a "native" record assembled out of history is
    // not a control at all, and under the payload refusal above it does not
    // even compile.
    const h = harness({ start: 1000, step: 7 });
    const nativeResult: SetPriorityResult = {
      outcome: "ok",
      tool: "setPriority",
      ticket: "4118",
      level: "High",
      reason: "customer escalated",
    };
    const native: readonly StepRecord[] = [
      {
        schemaVersion: SCHEMA_VERSION,
        now: 1000,
        sig: V1_SIG,
        staged: [],
        actions: [{ tool: "setPriority", input: { ticket: "4118", level: "High" } }],
        // SEALED, because a committed record only ever holds what the boundary
        // itself produced (spine/pure/tool-result). The control is still built
        // from a real v2 literal — the seal is the spine's stamp on it, not a
        // second shape.
        results: [seal(nativeResult)],
        commands: [],
        context: { promptVersion: "triage-prompt@1", digest: "" },
      },
    ];

    const replayed = refold(h.app.initial, native, h.app.dispatchers, h.app.licences);
    expect(replayed.effects.map((keyed) => keyed.effect)).toEqual([
      {
        kind: "LogDecision",
        at: 1000,
        effectClass: "Routine",
        ticket: "4118",
        level: "High",
        supersedes: "Normal",
        reason: "customer escalated",
      },
    ]);
  });

  it("a LIVE step commits the envelope, at the one site that mints a record", () => {
    const h = harness({ start: 1000, step: 7 });
    h.app.boundary.agent.submit({
      staged: [],
      actions: [{ tool: "setPriority", input: { ticket: "4118", level: "High" } }],
    });
    expect(h.app.bus.records().map((r) => r.schemaVersion)).toEqual([2]);
  });
});
